File: solutions/fsharp/space-age/1/SpaceAge.fs
module SpaceAge
open System

type Planet =
  | Earth
  | Mercury
  | Venus
  | Mars
  | Jupiter
  | Saturn
  | Uranus
  | Neptune

// let EarthYear = 31557600 |> decimal
let EarthYear = 1 |> decimal

let calculateYear (p : Planet) : decimal = 
  match p with 
  | Earth -> EarthYear
  | Mercury -> EarthYear * 0.2408467M
  | Venus -> EarthYear * 0.61519726M
  | Mars -> EarthYear * 1.8808158M
  | Jupiter -> EarthYear * 11.862615M
  | Saturn -> EarthYear * 29.447498M
  | Uranus -> EarthYear * 84.016846M
  | Neptune -> EarthYear * 164.79132M


let secToYear (s : decimal) = 
  s / 3600m / 24m / 365.25m

let spaceAge (planet: Planet) (seconds: decimal): decimal = 
  planet
  |> calculateYear
  |> (*) seconds
  |> secToYear

File: solutions/haskell/space-age/1/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune

secYears :: Float
secYears = 31557600.0

earthYear :: Planet -> Float
earthYear p =
    case p of 
        Mercury -> 0.2408467
        Venus   -> 0.61519726
        Earth   -> 1.0
        Mars    -> 1.8808158
        Jupiter -> 11.86261
        Saturn  -> 29.447498
        Uranus  -> 84.016846
        Neptune -> 164.79132

ageOn :: Planet -> Float -> Float
ageOn p sec = (sec / secYears) * (1/earthYear p) 

